web_ui/prometheus_test.go

```diff
@@ -0,0 +1,187 @@
+package web_ui
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/prometheus/common/route"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPrometheusProtection(t *testing.T) {
+
+	/*
+	* Tests that prometheus metrics are behind the origin's and federation's token. Specifically it signs a token
+	* with the origin's key and invokes a prometheus GET endpoint with both URL and Header authorization, with the
+	* URL authorization, it mimics matching the Federation URL to ensure that check is done, but intercepts with
+	* returning the origin jwk for testing purposes.
+	* This then does so again with an invalid token and confirms that the correct error is returned
+	 */
+
+	// Setup httptest recorder and context for the the unit test
+	viper.Reset()
+
+	av1 := route.New().WithPrefix("/api/v1.0/prometheus")
+
+	w := httptest.NewRecorder()
+	c, r := gin.CreateTestContext(w)
+	// Note, this handler function intercepts the "http.Get call to the federation uri
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		issuerKeyFile := viper.GetString("IssuerKey")
+		contents, err := os.ReadFile(issuerKeyFile)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = w.Write(contents)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}))
+	defer ts.Close()
+	c.Request = &http.Request{
+		URL: &url.URL{},
+	}
+
+	// Create temp dir for the origin key file
+	tDir := t.TempDir()
+	kfile := filepath.Join(tDir, "testKey")
+
+	//Setup a private key and a token
+	viper.Set("IssuerKey", kfile)
+
+	// Generate the origin private and public keys
+	_, err := config.LoadPublicKey("", kfile)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	privKey, err := config.LoadPrivateKey(kfile)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a token
+	issuerURL := url.URL{}
+	issuerURL.Scheme = "https"
+	issuerURL.Host = "test-http"
+	now := time.Now()
+	tok, err := jwt.NewBuilder().
```

Comment:
Moving forward, we should probably start making sure our internal tokens adhere to a well-known JWT profile. Brian indicated he favors the WLCG1 profile, so let's stick with that for now. You can find more information about the additional claims to include [here](https://github.com/WLCG-AuthZ-WG/common-jwt-profile/blob/master/profile.md#capability-based-authorization-scope)